File: scripts/data-conversion/flightJsonConverter/FlightHierarchyJSONWriter.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map.Entry;

import com.opencsv.CSVReader;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;

public class FlightHierarchyJSONWriter {
	static HashMap<Integer, String[]> zoom1 = new HashMap<Integer, String[]>();
	static HashMap<Integer, String[]> zoom2 = new HashMap<Integer, String[]>();
	static HashMap<Integer, String[]> zoom3 = new HashMap<Integer, String[]>();
	static HashMap<Integer, String[]> zoom4 = new HashMap<Integer, String[]>();
	static HashMap<Integer, String[]> zoom5 = new HashMap<Integer, String[]>();

	public static void main(String[] args) throws IOException {
		try {
			CSVReader reader = new CSVReader(new FileReader(
				"FlightroutesJava.csv"));
			String[] nextLine;
			while ((nextLine = reader.readNext()) != null) {

				double distance = calculateDistance(
						Double.parseDouble(nextLine[1]),
						Double.parseDouble(nextLine[2]),
						Double.parseDouble(nextLine[3]),
						Double.parseDouble(nextLine[4]));
				if (distance >= 5000) {
					zoom1.put(Integer.parseInt(nextLine[0]), nextLine);
				} else if (distance >= 3000 && distance < 5000) {
					zoom2.put(Integer.parseInt(nextLine[0]), nextLine);
				} else if (distance >= 1500 && distance < 3000) {
					zoom3.put(Integer.parseInt(nextLine[0]), nextLine);
				} else if (distance >= 500 && distance < 1500) {
					zoom4.put(Integer.parseInt(nextLine[0]), nextLine);
				} else if (distance >= 0 && distance < 500) {
					zoom5.put(Integer.parseInt(nextLine[0]), nextLine);
				}
			}
			reader.close();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		}

		defineParent(zoom2, zoom1);
		defineParent(zoom3, zoom2);
		defineParent(zoom4, zoom3);
		defineParent(zoom5, zoom4);

		JSONObject flightJSON = new JSONObject();
		JSONObject meta = new JSONObject();
		JSONObject root = new JSONObject();
		meta.put("temporal", new Boolean(false));
		meta.put("dataType", new String("flights"));
		flightJSON.put("meta", meta);

		JSONArray level1 = new JSONArray();

		for (Entry<Integer, String[]> entry1 : zoom1.entrySet()) {
			JSONObject flightObject1 = new JSONObject();
			JSONObject coordinatesSource1 = new JSONObject();
			JSONObject coordinatesDestination1 = new JSONObject();
			coordinatesSource1.put("latitude",
					new Double(Double.parseDouble(entry1.getValue()[1])));
			coordinatesSource1.put("longitude",
					new Double(Double.parseDouble(entry1.getValue()[2])));
			coordinatesDestination1.put("latitude",
					new Double(Double.parseDouble(entry1.getValue()[3])));
			coordinatesDestination1.put("longitude",
					new Double(Double.parseDouble(entry1.getValue()[4])));

			flightObject1.put("startposition", coordinatesSource1);
			flightObject1.put("endposition", coordinatesDestination1);
			// -------------------------------------------------------------------------------------------------
			JSONArray level2 = new JSONArray();
			for (Entry<Integer, String[]> entry2 : zoom2.entrySet()) {
				if (entry2.getValue()[6].equals(entry1.getValue()[0])) {
					JSONObject flightObject2 = new JSONObject();
					JSONObject coordinatesSource2 = new JSONObject();
					JSONObject coordinatesDestination2 = new JSONObject();
					coordinatesSource2
							.put("latitude",
									new Double(Double.parseDouble(entry2
											.getValue()[1])));
					coordinatesSource2
							.put("longitude",
									new Double(Double.parseDouble(entry2
											.getValue()[2])));
					coordinatesDestination2
							.put("latitude",
									new Double(Double.parseDouble(entry2
											.getValue()[3])));
					coordinatesDestination2
							.put("longitude",
									new Double(Double.parseDouble(entry2
											.getValue()[4])));

					flightObject2.put("startposition", coordinatesSource2);
					flightObject2.put("endposition", coordinatesDestination2);
					// -------------------------------------------------------------------------------------------------
					JSONArray level3 = new JSONArray();
					for (Entry<Integer, String[]> entry3 : zoom3.entrySet()) {
						if (entry3.getValue()[6].equals(entry2.getValue()[0])) {
							JSONObject flightObject3 = new JSONObject();
							JSONObject coordinatesSource3 = new JSONObject();
							JSONObject coordinatesDestination3 = new JSONObject();
							coordinatesSource3.put("latitude", new Double(
									Double.parseDouble(entry3.getValue()[1])));
							coordinatesSource3.put("longitude", new Double(
									Double.parseDouble(entry3.getValue()[2])));
							coordinatesDestination3.put("latitude", new Double(
									Double.parseDouble(entry3.getValue()[3])));
							coordinatesDestination3.put(
									"longitude",
									new Double(Double.parseDouble(entry3
											.getValue()[4])));

							flightObject3.put("startposition",
									coordinatesSource3);
							flightObject3.put("endposition",
									coordinatesDestination3);
							// -------------------------------------------------------------------------------------------------
							JSONArray level4 = new JSONArray();
							for (Entry<Integer, String[]> entry4 : zoom4
									.entrySet()) {
								if (entry4.getValue()[6].equals(entry3
										.getValue()[0])) {
									JSONObject flightObject4 = new JSONObject();
									JSONObject coordinatesSource4 = new JSONObject();
									JSONObject coordinatesDestination4 = new JSONObject();
									coordinatesSource4.put(
											"latitude",
											new Double(Double
													.parseDouble(entry4
															.getValue()[1])));
									coordinatesSource4.put(
											"longitude",
											new Double(Double
													.parseDouble(entry4
															.getValue()[2])));
									coordinatesDestination4.put(
											"latitude",
											new Double(Double
													.parseDouble(entry4
															.getValue()[3])));
									coordinatesDestination4.put(
											"longitude",
											new Double(Double
													.parseDouble(entry4
															.getValue()[4])));

									flightObject4.put("startposition",
											coordinatesSource4);
									flightObject4.put("endposition",
											coordinatesDestination4);
									// -------------------------------------------------------------------------------------------------
									JSONArray level5 = new JSONArray();
									for (Entry<Integer, String[]> entry5 : zoom5
											.entrySet()) {
										if (entry5.getValue()[6].equals(entry4
												.getValue()[0])) {
											JSONObject flightObject5 = new JSONObject();
											JSONObject coordinatesSource5 = new JSONObject();
											JSONObject coordinatesDestination5 = new JSONObject();
											coordinatesSource5
													.put("latitude",
															new Double(
																	Double.parseDouble(entry5
																			.getValue()[1])));
											coordinatesSource5
													.put("longitude",
															new Double(
																	Double.parseDouble(entry5
																			.getValue()[2])));
											coordinatesDestination5
													.put("latitude",
															new Double(
																	Double.parseDouble(entry5
																			.getValue()[3])));
											coordinatesDestination5
													.put("longitude",
															new Double(
																	Double.parseDouble(entry5
																			.getValue()[4])));

											flightObject5.put("startposition",
													coordinatesSource5);
											flightObject5.put("endposition",
													coordinatesDestination5);

											level5.add(flightObject5);
										}
									}
									flightObject4.put("children", level5);
									// -------------------------------------------------------------------------------------------------
									level4.add(flightObject4);
								}
							}
							flightObject3.put("children", level4);
							// -------------------------------------------------------------------------------------------------
							level3.add(flightObject3);
						}
					}
					flightObject2.put("children", level3);
					// -------------------------------------------------------------------------------------------------
					level2.add(flightObject2);
				}
			}
			flightObject1.put("children", level2);
			// -------------------------------------------------------------------------------------------------
			level1.add(flightObject1);

		}
		root.put("children", level1);
		flightJSON.put("root", root);
		writeJSONFile(flightJSON);
	}

	public static int checkMinDistance(HashMap<Integer, String[]> zoomLevel,
			double lat, double lon) {
		int minIndex = 0;
		double minDistance = Double.MAX_VALUE;
		for (Entry<Integer, String[]> entry : zoomLevel.entrySet()) {
			double distance = calculateDistance(lat, lon,
					Double.parseDouble(entry.getValue()[1]),
					Double.parseDouble(entry.getValue()[2]));
			if (distance < minDistance) {
				minDistance = distance;
				minIndex = Integer.parseInt(entry.getValue()[0]);
			}
		}

		return minIndex;
	}

	public static double calculateDistance(double lat1, double lon1,
			double lat2, double lon2) {
		double lat = ((lat1 + lat2) / 2 * (Math.PI / 180.0));
		double widthDegreeLat = 111.3;
		double widthDegreeLon = widthDegreeLat * Math.cos(lat);
		double dx = widthDegreeLon * (lon1 - lon2);
		double dy = widthDegreeLat * (lat1 - lat2);
		// checks if fligth is over map-border and subtracts
		// the distance from the earths circumference (approx. 40074km)
		double rawDistance = Math.sqrt(dx * dx + dy * dy);
		if (rawDistance > 20037) {
			return 40074 - rawDistance;
		} else {
			return rawDistance;
		}

	}

	public static void defineParent(HashMap<Integer, String[]> childMap,
			HashMap<Integer, String[]> parentMap) {
		for (Entry<Integer, String[]> entry : childMap.entrySet()) {
			int parentIndex = checkMinDistance(parentMap,
					Double.parseDouble(entry.getValue()[1]),
					Double.parseDouble(entry.getValue()[2]));
			entry.getValue()[6] = Integer.toString(parentIndex);
		}
	}

	public static void writeJSONFile(JSONObject flightsObject) {
		try {
			File jsonFile = new File("flights.json");
			FileWriter writer = new FileWriter(jsonFile);
			writer.write(flightsObject.toJSONString());
			writer.flush();
			writer.close();
			System.out.println("File succesfully converted");
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}

File: scripts/data-conversion/flightJsonConverter/FlightHierarchyJSONWriter.py
"""
skript uses python 3.5
"""
import json
import sys
import csv
import math
from DataUtility import DataUtility

def main(argv):

    zoom1 = []
    zoom2 = []
    zoom3 = []
    zoom4 = []
    zoom5 = []
	
    d = DataUtility()
    csv_data = d.read_csv_file('Flightroutes.csv')
    #fills the 5 zoomlists which represents the priority-levels 
    for row in csv_data:
        distance = calculateDistance(float(row["srcLatitude"]), float(row["srcLongitude"]),float(row["destLatitude"]),float(row["destLongitude"]))
        flight = [int(row["id"]),float(row["srcLatitude"]),float(row["srcLongitude"]),float(row["destLatitude"]),float(row["destLongitude"]),int(row["parent"])]
        if distance >= 5000:       
            zoom1.append(flight)
        elif(distance < 5000 and distance >= 3000):
            zoom2.append(flight)
        elif(distance < 3000 and distance >= 1500):
            zoom3.append(flight)
        elif(distance < 1500 and distance >= 500):
            zoom4.append(flight)
        elif(distance < 500):
            zoom5.append(flight)
    #defines the child-parent relationship for every flight 
    defineParent(zoom2, zoom1)
    defineParent(zoom3, zoom2)
    defineParent(zoom4, zoom3)
    defineParent(zoom5, zoom4)
    counter = 0
    level1Entries = []
    for entry1 in zoom1:
#################################################### start level 2
        level2Entries = []
        for entry2 in zoom2:
#################################################### start level 3            
            level3Entries = []
            for entry3 in zoom3:
#################################################### start level 4                            
                level4Entries = []
                for entry4 in zoom4:
#################################################### start level 5                            
                    level5Entries = []
                    for entry5 in zoom5:
                        if entry4[0] == entry5[5]:          
                            flight = {'startPosition':{'latitude':entry5[1],'longitude':entry5[2]},'endPosition':{'latitude':entry5[3],'longitude':entry5[4]}} 
                            level5Entries.append(flight)
#################################################### end level 5  
                    if entry3[0] == entry4[5]:          
                        flight = {'startPosition':{'latitude':entry4[1],'longitude':entry4[2]},'endPosition':{'latitude':entry4[3],'longitude':entry4[4]},'children':level5Entries} 
                        level4Entries.append(flight)
#################################################### end level 4                 
                if entry2[0] == entry3[5]:
                    flight = {'startPosition':{'latitude':entry3[1],'longitude':entry3[2]},'endPosition':{'latitude':entry3[3],'longitude':entry3[4]},'children':level4Entries}
                    level3Entries.append(flight)   
#################################################### end level 3            
            if entry1[0] == entry2[5]:
                flight = {'startPosition':{'latitude':entry2[1],'longitude':entry2[2]},'endPosition':{'latitude':entry2[3],'longitude':entry2[4]},'children':level3Entries}            
                level2Entries.append(flight)
#################################################### end level 2       
        flight = {'startPosition':{'latitude':entry1[1],'longitude':entry1[2]},'endPosition':{'latitude':entry1[3],'longitude':entry1[4]},'children':level2Entries}
        level1Entries.append(flight)
        level1Entries.append(flight)

    flights = {'meta': {'dataType': 'flights', 'temporal': False}, 'root': {'children': level1Entries}}	
    d.write_json_file(flights,'flights.json', True)
	
"""
defines the hierarchical parent of a flight and writes it in to 
the flights array 
"""
def defineParent(childList, parentList):
    for entry in childList:
        parentIndex = checkMindistance(parentList, entry[1], entry[2])
        entry[5] = parentIndex
		
"""
approximation of distance betwenn two coordinates in kilometers.
an approximation is enough in this case, because it only results 
in one of five priority-levels
"""
def calculateDistance(lat1, lon1, lat2, lon2):
    lat = ((lat1 + lat2) / (math.pi/180))
    widthDegreeLat = 111.3
    widthDegreeLon = widthDegreeLat * math.cos(lat)
    dx = widthDegreeLon * (lon1 - lon2)
    dy = widthDegreeLat * (lat1 - lat2)
    # checks if fligth is over map-border and subtracts
    # the distance from the earths circumference (approx. 40074km)
    rawdistance = math.sqrt(dx * dx + dy * dy)
    if (rawdistance > 20037):
        return 40074-rawdistance
    else:
        return rawdistance	
	
"""
returns the index of the parentFlight, which has the minimum distance between
the source-airports
"""
def checkMindistance(zoomlevel, lat, lon):
    minIndex = 0
    minDistance = 50000
    for entry in zoomlevel:
        distance = calculateDistance(lat, lon, entry[1], entry[2])
        if distance < minDistance:
            minDistance = distance
            minIndex = entry[0]
    return minIndex
		
if __name__ == "__main__":
    main(sys.argv[1:])